modules/system/etc.nix

```diff
@@ -42,64 +39,103 @@ in
+            etcFileSha256Hash=''${etcFileSha256Output%% *}
+            for knownSha256Hash in ''${etcSha256Hashes[$subPath]}; do
+              if [[ $etcFileSha256Hash == $knownSha256Hash ]]; then
+                # Hash matches, OK to overwrite; go to the next file.
+                continue 2
+              fi
+            done
+          fi
+          etcProblems+=("$etcFile")
+        fi
+      done < <(find -H "$systemConfig/etc" -type l -print0)
+
+      if (( ''${#etcProblems[@]} )); then
+        printf >&2 '\x1B[1;31merror: Unexpected files in /etc, aborting '
+        printf >&2 'activation\x1B[0m\n'
+        printf >&2 'The following files have unrecognized content and would be '
+        printf >&2 'overwritten:\n\n'
+        printf >&2 '  %s\n' "''${etcProblems[@]}"
+        printf >&2 '\nPlease check there is nothing critical in these files, '
+        printf >&2 'rename them by adding .before-nix-darwin to the end, and '
+        printf >&2 'then try again.\n'
+        exit 2
+      fi
+    '';
+
     system.activationScripts.etc.text = ''
       # Set up the statically computed bits of /etc.
-      echo "setting up /etc..." >&2
-
-      declare -A etcSha256Hashes
-      ${concatMapStringsSep "\n" (attr: "etcSha256Hashes['/etc/${attr.target}']='${concatStringsSep " " attr.knownSha256Hashes}'") etc}
+      printf >&2 'setting up /etc...\n'
 
       ln -sfn "$(readlink -f $systemConfig/etc)" /etc/static
 
-      errorOccurred=false
-      for f in $(find /etc/static/* -type l); do
-        l=/etc/''${f#/etc/static/}
-        d=''${l%/*}
-        if [ ! -e "$d" ]; then
-          mkdir -p "$d"
+      while IFS= read -r -d "" etcStaticFile; do
+        etcFile=/etc/''${etcStaticFile#/etc/static/}
+        etcDir=''${etcFile%/*}
+
+        if [[ ! -d $etcDir ]]; then
+          mkdir -p "$etcDir"
         fi
-        if [ -e "$f".copy ]; then
-          cp "$f" "$l"
+
+        if [[ -e $etcStaticFile.copy ]]; then
+          cp "$etcStaticFile" "$etcFile"
           continue
         fi
-        if [ -e "$l" ]; then
-          if [ "$(readlink "$l")" != "$f" ]; then
-            if ! grep -q /etc/static "$l"; then
-              o=''$(shasum -a256 "$l")
-              o=''${o%% *}
-              for h in ''${etcSha256Hashes["$l"]}; do
-                if [ "$o" = "$h" ]; then
-                  mv "$l" "$l.before-nix-darwin"
-                  ln -s "$f" "$l"
-                  break
-                else
-                  h=
-                fi
-              done
-
-              if [ -z "$h" ]; then
-                echo "[1;31merror: not linking environment.etc.\"''${l#/etc/}\" because $l already exists, skipping...[0m" >&2
-                echo "[1;31mexisting file has unknown content $o, move and activate again to apply[0m" >&2
-                errorOccurred=true
-              fi
-            fi
+
+        if [[ -e $etcFile ]]; then
+          if [[ $(readlink "$etcFile") == $etcStaticFile ]]; then
```

Comment:
```suggestion
          if [[ $(readlink "$etcFile") == "$etcStaticFile" ]]; then
```

```
    if [[ $(readlink "$etcFile") == $etcStaticFile ]]; then
                                    ^------------^ SC2053 (warning): Quote the right-hand side of == in [[ ]] to prevent glob matching.
```